# **소설 작가 분류 AI 경진대회 ** - 데이콘

https://dacon.io/competitions/official/235670/overview/description

[주제] 

**CSV 파일의 'text'를 분석해서 'author' 예측**

[과정]
* pandas를 통해 읽어와 train set과 test set으으로 분리(3:1)
* train set으로 학습하고, test set으로 평가

[제출 내용]
* 머신러닝 기법 중 적절한 것들을 선택해 제일 좋은 모형 찾기
* train set과 test set에 대한 평가

<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#1.-데이터-준비" data-toc-modified-id="1.-데이터-준비-1"><span class="toc-item-num">1&nbsp;&nbsp;</span><strong>1. 데이터 준비</strong></a></span><ul class="toc-item"><li><span><a href="#1.1-train-set과-test-set-분할(비율-3:1)" data-toc-modified-id="1.1-train-set과-test-set-분할(비율-3:1)-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>1.1 train set과 test set 분할(비율 3:1)</a></span></li><li><span><a href="#1.2-데이터-확인" data-toc-modified-id="1.2-데이터-확인-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>1.2 데이터 확인</a></span></li><li><span><a href="#1.3-정규표현식을-이용한-토크나이저-함수-정의" data-toc-modified-id="1.3-정규표현식을-이용한-토크나이저-함수-정의-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span><strong>1.3 정규표현식을 이용한 토크나이저 함수 정의</strong></a></span></li><li><span><a href="#1.4-TfidfVectorizer에-토큰화-결과를-입력으로-사용" data-toc-modified-id="1.4-TfidfVectorizer에-토큰화-결과를-입력으로-사용-1.4"><span class="toc-item-num">1.4&nbsp;&nbsp;</span><strong>1.4 TfidfVectorizer에 토큰화 결과를 입력으로 사용</strong></a></span></li><li><span><a href="#1.5-첫-번째-문서에-대해서-200개만-확인" data-toc-modified-id="1.5-첫-번째-문서에-대해서-200개만-확인-1.5"><span class="toc-item-num">1.5&nbsp;&nbsp;</span>1.5 첫 번째 문서에 대해서 200개만 확인</a></span></li></ul></li><li><span><a href="#2.-문서-분류-알고리즘에-적용" data-toc-modified-id="2.-문서-분류-알고리즘에-적용-2"><span class="toc-item-num">2&nbsp;&nbsp;</span><strong>2. 문서 분류 알고리즘에 적용</strong></a></span><ul class="toc-item"><li><span><a href="#2.1-나이브-베이즈-분류기-이용" data-toc-modified-id="2.1-나이브-베이즈-분류기-이용-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>2.1 나이브 베이즈 분류기 이용</a></span></li><li><span><a href="#2.2-로지스틱-회귀분석-이용" data-toc-modified-id="2.2-로지스틱-회귀분석-이용-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>2.2 로지스틱 회귀분석 이용</a></span></li><li><span><a href="#2.3-릿지-회귀-이용" data-toc-modified-id="2.3-릿지-회귀-이용-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>2.3 릿지 회귀 이용</a></span></li><li><span><a href="#2.4-라쏘-회귀-이용" data-toc-modified-id="2.4-라쏘-회귀-이용-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>2.4 라쏘 회귀 이용</a></span></li><li><span><a href="#2.5-결정-트리-이용" data-toc-modified-id="2.5-결정-트리-이용-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>2.5 결정 트리 이용</a></span></li><li><span><a href="#2.6-랜덤-포레스트-이용" data-toc-modified-id="2.6-랜덤-포레스트-이용-2.6"><span class="toc-item-num">2.6&nbsp;&nbsp;</span>2.6 랜덤 포레스트 이용</a></span></li><li><span><a href="#2.7-그래디언트-부스팅-이용" data-toc-modified-id="2.7-그래디언트-부스팅-이용-2.7"><span class="toc-item-num">2.7&nbsp;&nbsp;</span>2.7 그래디언트 부스팅 이용</a></span></li></ul></li><li><span><a href="#3.-종합적인-결과-보기" data-toc-modified-id="3.-종합적인-결과-보기-3"><span class="toc-item-num">3&nbsp;&nbsp;</span><strong>3. 종합적인 결과 보기</strong></a></span></li><li><span><a href="#4.-test-set에-대해서-분류-예측" data-toc-modified-id="4.-test-set에-대해서-분류-예측-4"><span class="toc-item-num">4&nbsp;&nbsp;</span><strong>4. test set에 대해서 분류 예측</strong></a></span></li><li><span><a href="#5.-author-별로-영향을-많이-미친-특성-혹은-단어-확인" data-toc-modified-id="5.-author-별로-영향을-많이-미친-특성-혹은-단어-확인-5"><span class="toc-item-num">5&nbsp;&nbsp;</span><strong>5. author 별로 영향을 많이 미친 특성 혹은 단어 확인</strong></a></span></li><li><span><a href="#6.-결론" data-toc-modified-id="6.-결론-6"><span class="toc-item-num">6&nbsp;&nbsp;</span><strong>6. 결론</strong></a></span></li></ul></div>

## **1. 데이터 준비**

In [1]:
import pandas as pd

df = pd.read_csv("train.csv")
df.head(10,)

index                                               text  author
0      0  He was almost choking. There was so much, so m...       3
1      1             “Your sister asked for it, I suppose?”       2
2      2   She was engaged one day as she walked, in per...       1
3      3  The captain was in the porch, keeping himself ...       4
4      4  “Have mercy, gentlemen!” odin flung up his han...       3
5      5  "It was well fought," he said, "and, by my soo...       4
6      6  Not to pay him was impossible, considering his...       3
7      7  “A proper figure of a man at-arms,” said the l...       2
8      8    'You were not here last Sunday night,' he said.       0
9      9  “You must not ask me that!” I cried. “Hell may...       4

### 1.1 train set과 test set 분할(비율 3:1)

In [2]:
from sklearn.model_selection import train_test_split

x_train, x_test, y_train, y_test = train_test_split(df.text, df.author, test_size=0.25, random_state=1)

print('# Train set size : ', len(x_train))
print('# Test set size : ', len(x_test))

# Train set size :  41159
# Test set size :  13720


### 1.2 데이터 확인

In [3]:
print('### Train set ###')
print('# text sample : ', x_train.head())
print('# author sample : ', y_train.head())
print()

print('### Test set ###')
print('# text sample : ', x_test.head())
print('# author sample : ', y_test.head())

### Train set ###
# text sample :  34509     “And then you have added so much to it yourse...
15230    Presently, however, the ground began to rise, ...
51072    Then followed an obstinate and deadly struggle...
11046    “Is a bit thick, you mean? Well, in a way that...
52441    'I hear you,' replied the girl, turning her fa...
Name: text, dtype: object
# author sample :  34509    1
15230    4
51072    4
11046    3
52441    0
Name: author, dtype: int64

### Test set ###
# text sample :  41956    “Ah, that’s true! I understand that perfectly ...
42056     He took the boots from the bag, and compared ...
16422    "It's not only absurd, but something else as w...
22151    "You're an atheist because you're a snob, a sn...
8409     I would not, however, have dwelt on such trivi...
Name: text, dtype: object
# author sample :  41956    3
42056    2
16422    3
22151    3
8409     3
Name: author, dtype: int64


### **1.3 정규표현식을 이용한 토크나이저 함수 정의**

In [4]:
from nltk.corpus import stopwords
from nltk.tokenize import RegexpTokenizer
from nltk.stem.porter import PorterStemmer
import re

In [5]:
RegTok = RegexpTokenizer("[\w']{3,}") # 정규표현식 이용

stop_words = set(stopwords.words('english')) # 영어 불용어 

def tokenizer(text):
    tokens = RegTok.tokenize(text.lower())
    words = [word for word in tokens if (word not in stop_words) and len(word) > 2] # 불용어 제거
    features = (list(map(lambda token:PorterStemmer().stem(token), words))) # 포터 스테머 적용
    return features

### **1.4 TfidfVectorizer에 토큰화 결과를 입력으로 사용**

In [ ]:
from sklearn.feature_extraction.text import TfidfVectorizer

tfidf = TfidfVectorizer(tokenizer=tokenizer).fit(x_train)

x_train_tfidf = tfidf.fit_transform(x_train)
x_test_tfidf = tfidf.transform(x_test)

### 1.5 첫 번째 문서에 대해서 200개만 확인

In [7]:
for word, count in zip(tfidf.get_feature_names()[:200], x_train_tfidf[0].toarray()[0, :200]):
    print(word, ':', count, end=', ')

'80' : 0.0, '_c'est : 0.0, '_i_ : 0.0, '_non : 0.0, '_real : 0.0, '_that_ : 0.0, '_there_ : 0.0, '_was_ : 0.0, '_your_ : 0.0, 'a : 0.0, 'a' : 0.0, 'about : 0.0, 'add : 0.0, 'address : 0.0, 'affect : 0.0, 'afraid : 0.0, 'after : 0.0, 'again : 0.0, 'ah : 0.0, 'aha : 0.0, 'ain't : 0.0, 'all : 0.0, 'alway : 0.0, 'am : 0.0, 'amen : 0.0, 'amiabl : 0.0, 'among : 0.0, 'an : 0.0, 'an't : 0.0, 'and : 0.0, 'ani : 0.0, 'anoth : 0.0, 'antimoni : 0.0, 'anyth : 0.0, 'apocalypse' : 0.0, 'appears' : 0.0, 'are : 0.0, 'aro : 0.0, 'ask : 0.0, 'at : 0.0, 'attend : 0.0, 'auld : 0.0, 'ave : 0.0, 'aweel : 0.0, 'ay : 0.0, 'ay' : 0.0, 'back : 0.0, 'bah : 0.0, 'bayton : 0.0, 'be : 0.0, 'bear : 0.0, 'beauti : 0.0, 'becaus : 0.0, 'befor : 0.0, 'bet : 0.0, 'better : 0.0, 'beyond : 0.0, 'bi : 0.0, 'bill : 0.0, 'bister : 0.0, 'black : 0.0, 'blather : 0.0, 'bless : 0.0, 'blood : 0.0, 'blow : 0.0, 'bolter : 0.0, 'bombarded' : 0.0, 'bout : 0.0, 'bow : 0.0, 'box' : 0.0, 'boy : 0.0, 'boy' : 0.0, 'brandish : 0.0, 'brass : 

## **2. 문서 분류 알고리즘에 적용**

### 2.1 나이브 베이즈 분류기 이용

In [8]:
from sklearn.naive_bayes import MultinomialNB

nb_clf = MultinomialNB()

nb_clf.fit(x_train_tfidf, y_train)

print('# Train set score : {:.3f}'. format(nb_clf.score(x_train_tfidf, y_train)))
print('# Test set score : {:.3f}'.format(nb_clf.score(x_test_tfidf, y_test)))

# Train set score : 0.734
# Test set score : 0.675


In [9]:
print('### alpha=0.01 ###')
nb_clf = MultinomialNB(alpha=0.01)
nb_clf.fit(x_train_tfidf, y_train)
print('# Train set score : {:.3f}'. format(nb_clf.score(x_train_tfidf, y_train)))
print('# Test set score : {:.3f}'.format(nb_clf.score(x_test_tfidf, y_test)))
print()

print('### alpha=0.05 ###')
nb_clf = MultinomialNB(alpha=0.05)
nb_clf.fit(x_train_tfidf, y_train)
print('# Train set score : {:.3f}'. format(nb_clf.score(x_train_tfidf, y_train)))
print('# Test set score : {:.3f}'.format(nb_clf.score(x_test_tfidf, y_test)))
print()

print('### alpha=1 ###')
nb_clf = MultinomialNB(alpha=1)
nb_clf.fit(x_train_tfidf, y_train)
print('# Train set score : {:.3f}'. format(nb_clf.score(x_train_tfidf, y_train)))
print('# Test set score : {:.3f}'.format(nb_clf.score(x_test_tfidf, y_test)))
print()

print('### alpha=1.5 ###')
nb_clf = MultinomialNB(alpha=1.5)
nb_clf.fit(x_train_tfidf, y_train)
print('# Train set score : {:.3f}'. format(nb_clf.score(x_train_tfidf, y_train)))
print('# Test set score : {:.3f}'.format(nb_clf.score(x_test_tfidf, y_test)))
print()

### alpha=0.01 ###
# Train set score : 0.811
# Test set score : 0.720

### alpha=0.05 ###
# Train set score : 0.807
# Test set score : 0.722

### alpha=1 ###
# Train set score : 0.734
# Test set score : 0.675

### alpha=1.5 ###
# Train set score : 0.701
# Test set score : 0.648



***alpha가 0.01일 때 가장 높은 정확도를 보인다.***

### 2.2 로지스틱 회귀분석 이용

In [12]:
from sklearn.linear_model import LogisticRegression

lr_clf = LogisticRegression()

lr_clf.fit(x_train_tfidf, y_train)

print('# Train set score : {:.3f}'.format(lr_clf.score(x_train_tfidf, y_train)))
print('# Test set score : {:.3f}'.format(lr_clf.score(x_test_tfidf, y_test)))

# Train set score : 0.787
# Test set score : 0.713


### 2.3 릿지 회귀 이용

In [13]:
from sklearn.linear_model import RidgeClassifier

ridge = RidgeClassifier() # alpha=1

ridge.fit(x_train_tfidf, y_train)

print('# Train set score : {:.3f}'.format(ridge.score(x_train_tfidf, y_train)))
print('# Test set score : {:.3f}'.format(ridge.score(x_test_tfidf, y_test)))

# Train set score : 0.842
# Test set score : 0.727


In [14]:
print('### alpha=1.2 ###')
ridge = RidgeClassifier(alpha=1.2)
ridge.fit(x_train_tfidf, y_train)
print('# Train set score : {:.3f}'.format(ridge.score(x_train_tfidf, y_train)))
print('# Test set score : {:.3f}'.format(ridge.score(x_test_tfidf, y_test)))
print()

print('### alpha=1.5 ###')
ridge = RidgeClassifier(alpha=1.5)
ridge.fit(x_train_tfidf, y_train)
print('# Train set score : {:.3f}'.format(ridge.score(x_train_tfidf, y_train)))
print('# Test set score : {:.3f}'.format(ridge.score(x_test_tfidf, y_test)))
print()

print('### alpha=2 ###')
ridge = RidgeClassifier(alpha=2)
ridge.fit(x_train_tfidf, y_train)
print('# Train set score : {:.3f}'.format(ridge.score(x_train_tfidf, y_train)))
print('# Test set score : {:.3f}'.format(ridge.score(x_test_tfidf, y_test)))
print()

print('### alpha=2.5 ###')
ridge = RidgeClassifier(alpha=2.5)
ridge.fit(x_train_tfidf, y_train)
print('# Train set score : {:.3f}'.format(ridge.score(x_train_tfidf, y_train)))
print('# Test set score : {:.3f}'.format(ridge.score(x_test_tfidf, y_test)))
print()

print('### alpha=3 ###')
ridge = RidgeClassifier(alpha=3)
ridge.fit(x_train_tfidf, y_train)
print('# Train set score : {:.3f}'.format(ridge.score(x_train_tfidf, y_train)))
print('# Test set score : {:.3f}'.format(ridge.score(x_test_tfidf, y_test)))
print()

### alpha=1.2 ###
# Train set score : 0.839
# Test set score : 0.727

### alpha=1.5 ###
# Train set score : 0.834
# Test set score : 0.726

### alpha=2 ###
# Train set score : 0.826
# Test set score : 0.725

### alpha=2.5 ###
# Train set score : 0.820
# Test set score : 0.724

### alpha=3 ###
# Train set score : 0.814
# Test set score : 0.723



***릿지 회귀에서 최적의 alpha는 기본 값인 1에서 train set에서는 84,2%, test set에서는 72.7%로 정확성이 가장 높다.***

### 2.4 라쏘 회귀 이용

In [15]:
lasso = LogisticRegression(penalty='l1', solver='liblinear', C=1)

lasso.fit(x_train_tfidf, y_train)

print('# Train set score : {:.3f}'.format(lasso.score(x_train_tfidf, y_train)))
print('# Test set score : {:.3f}'.format(lasso.score(x_test_tfidf, y_test)))

# Train set score : 0.755
# Test set score : 0.696


In [16]:
print('### C = 0.8 ###')
lasso = LogisticRegression(penalty='l1', solver='liblinear', C=0.8)
lasso.fit(x_train_tfidf, y_train)
print('# Train set score : {:.3f}'.format(lasso.score(x_train_tfidf, y_train)))
print('# Test set score : {:.3f}'.format(lasso.score(x_test_tfidf, y_test)))
print()

print('### C = 0.5 ###')
lasso = LogisticRegression(penalty='l1', solver='liblinear', C=0.5)
lasso.fit(x_train_tfidf, y_train)
print('# Train set score : {:.3f}'.format(lasso.score(x_train_tfidf, y_train)))
print('# Test set score : {:.3f}'.format(lasso.score(x_test_tfidf, y_test)))
print()

print('### C = 0.1 ###')
lasso = LogisticRegression(penalty='l1', solver='liblinear', C=0.1)
lasso.fit(x_train_tfidf, y_train)
print('# Train set score : {:.3f}'.format(lasso.score(x_train_tfidf, y_train)))
print('# Test set score : {:.3f}'.format(lasso.score(x_test_tfidf, y_test)))
print()

print('### C = 0.01 ###')
lasso = LogisticRegression(penalty='l1', solver='liblinear', C=0.01)
lasso.fit(x_train_tfidf, y_train)
print('# Train set score : {:.3f}'.format(lasso.score(x_train_tfidf, y_train)))
print('# Test set score : {:.3f}'.format(lasso.score(x_test_tfidf, y_test)))
print()

### C = 0.8 ###
# Train set score : 0.736
# Test set score : 0.687

### C = 0.5 ###
# Train set score : 0.695
# Test set score : 0.660

### C = 0.1 ###
# Train set score : 0.530
# Test set score : 0.527

### C = 0.01 ###
# Train set score : 0.318
# Test set score : 0.321



***라쏘 회귀에서 C가 1일 때 train set에서는 75.5%, test set에서는 69.6%의 정확도로 가장 높다.***

### 2.5 결정 트리 이용

In [17]:
from sklearn.tree import DecisionTreeClassifier

dec_tree = DecisionTreeClassifier(random_state=1)

dec_tree.fit(x_train_tfidf, y_train)

print('# Train set score : {:.3f}'.format(dec_tree.score(x_train_tfidf, y_train)))
print('# Test set score : {:.3f}'.format(dec_tree.score(x_test_tfidf, y_test)))

# Train set score : 0.994
# Test set score : 0.444


train set에 대해서는 99.4%인 반면 test set에 대해서는 44.4%로 정확도에 큰 차이가 있어 과적합 되었다고 할 수 있다.

### 2.6 랜덤 포레스트 이용

In [18]:
from sklearn.ensemble import RandomForestClassifier

forest = RandomForestClassifier(random_state=1)

forest.fit(x_train_tfidf, y_train)

print('# Train set score : {:.3f}'.format(forest.score(x_train_tfidf, y_train)))
print('# Test set score : {:.3f}'.format(forest.score(x_test_tfidf, y_test)))

# Train set score : 0.984
# Test set score : 0.519


랜덤 포레스트를 사용했을 때도 마찬가지로 결정트리와 비슷하게 train set에 대해서는 정확도가 높은 반면에 test set에 대해서는 훨씬 낮기 때문에 과적합 되었다고 할 수 있다.

### 2.7 그래디언트 부스팅 이용

In [19]:
from sklearn.ensemble import GradientBoostingClassifier

gb_clf = GradientBoostingClassifier(random_state=1)

gb_clf.fit(x_train_tfidf, y_train)

print('# Train set score : {:.3f}'.format(gb_clf.score(x_train_tfidf, y_train)))
print('# Test set score : {:.3f}'.format(gb_clf.score(x_test_tfidf, y_test)))

# Train set score : 0.565
# Test set score : 0.540


train set에 대해서는 56.5%, test set에 대해서는 54.0%의 정확도로 과적합 되었다고 판단되지는 않지만 정확도가 떨어진다.

## **3. 종합적인 결과 보기**

**(결정트리, 랜덤 포레스트, 그래디언트 부스팅 제외)**

In [20]:
print('### Naive_Bayes Classifier ###') # alpha가 0.01일 때 가장 높은 정확도 
nb_clf = MultinomialNB(alpha=0.01)
nb_clf.fit(x_train_tfidf, y_train)
print('# Train set score : {:.3f}'. format(nb_clf.score(x_train_tfidf, y_train)))
print('# Test set score : {:.3f}'.format(nb_clf.score(x_test_tfidf, y_test)))
print()

print('### Logistic_Regression Classifier ###')
print('# Train set score : {:.3f}'.format(lr_clf.score(x_train_tfidf, y_train)))
print('# Test set score : {:.3f}'.format(lr_clf.score(x_test_tfidf, y_test)))
print()

print('### Ridge Classifier ###') # alpha=1일 떄 가장 높은 정확도
ridge = RidgeClassifier() 
ridge.fit(x_train_tfidf, y_train)
print('# Train set score : {:.3f}'.format(ridge.score(x_train_tfidf, y_train)))
print('# Test set score : {:.3f}'.format(ridge.score(x_test_tfidf, y_test)))
print()

print('### Lasso Classifier ###') # C가 1일 때 가장 높은 정확도
lasso = LogisticRegression(penalty='l1', solver='liblinear', C=1)
lasso.fit(x_train_tfidf, y_train)
print('# Train set score : {:.3f}'.format(lasso.score(x_train_tfidf, y_train)))
print('# Test set score : {:.3f}'.format(lasso.score(x_test_tfidf, y_test)))

### Naive_Bayes Classifier ###
# Train set score : 0.811
# Test set score : 0.720

### Logistic_Regression Classifier ###
# Train set score : 0.787
# Test set score : 0.713

### Ridge Classifier ###
# Train set score : 0.842
# Test set score : 0.727

### Lasso Classifier ###
# Train set score : 0.755
# Test set score : 0.696


***릿지 회귀를 이용한 분류기의 경우가 train set에서는 83.4%, test set에 대해서는 72.6%로 가장 높은 정확도를 갖는다.***

## **4. test set에 대해서 분류 예측**

In [143]:
pd.concat([x_test[:10], y_test[:10]], axis=1)

text  author
41956  “Ah, that’s true! I understand that perfectly ...       3
42056   He took the boots from the bag, and compared ...       2
16422  "It's not only absurd, but something else as w...       3
22151  "You're an atheist because you're a snob, a sn...       3
8409   I would not, however, have dwelt on such trivi...       3
46026  “Mr. Hands,” he said, “here are two of us with...       4
85        “We'll never get ashore at this rate,” said I.       4
45368  “And now speak to your father, dearest. No oth...       0
19300  As he spoke, a French squire and the Bohemian ...       2
41708  If I were still not brought to reason by all t...       3

In [144]:
pred_head = ridge.predict(x_test_tfidf[:10])

print('# author 분류 결과 : ', pred_head)

# author 분류 결과 :  [1 2 3 3 3 2 4 0 2 3]


***앞의 10개의 소설 중 8개는 올바르게 분류되었고 2개는 제대로 분류되지 않았다.***

In [157]:
pd.concat([x_test[-10:], y_test[-10:]], axis=1)

text  author
37943      'Conkey means Nosey, ma'am,' interposed Duff.       0
38349   Every object in the next day’s journey was ne...       1
10029  “You may smile,--but there’s a career in this,...       3
28874   "Swear to me that he still lives," she returned.       4
21086  ‘Aye, aye?’ said odin, returning. ‘I am glad t...       0
30720  "Oh, that's the best explanation you can give,...       2
45434  'Know me!' cried odin. 'Who can do so? My life...       4
5848   "We shall fight at daybreak, that's a settled ...       3
1616   With that, odin, who was flushed with walking,...       0
26233  “Can ye so, indeed?” asked the Jacobite. “Well...       4

In [158]:
pred_tail = ridge.predict(x_test_tfidf[-10:])

print('# author 분류 결과 : ', pred_tail)

# author 분류 결과 :  [0 1 3 4 4 3 3 3 0 4]


***뒤의 10개의 소설 중에서도 7개는 올바르게 분류되었지만 3개는 잘못 분류 되었다.***

## **5. author 별로 영향을 많이 미친 특성 혹은 단어 확인**

In [159]:
import numpy as np

def top_features(classifier, vectorizer, categories, n):
    feature_names = np.asarray(vectorizer.get_feature_names())
    
    for i, category in enumerate(categories):
        top = np.argsort(-classifier.coef_[i])[:n]
        print('%s: %s' % (category, ', '.join(feature_names[top])))

In [165]:
authors = [0, 1, 2, 3, 4]

In [166]:
top_features(ridge, tfidf, authors, 10) # 상위 10개의 단어 확인

0: dora, darnay, joe, aunt, jerri, carton, handel, pip, creakl, crupp
1: ferrier, luci, lizzi, ann, shew, surpriz, tilney, jane, margaret, maria
2: inspector, professor, hordl, nigel, hudson, moor, bork, bowman, straker, pinner
3: princ, roubl, stepan, totski, moscow, ivan, suddenli, podincutor, simpli, zossimov
4: sila, casimir, villon, florizel, stubb, countess, silver, marjori, pool, balli


## **6. 결론**

**릿지 회귀**에 데이터를 적용했을 때 **train set에서는 84.2%, test set에서는 72.7%의 정확도**로 나이브 베이즈 분류기, 로지스틱 회귀, 라쏘 회귀 등에 적용했을 때보다 정확성이 높게 나타난다. 결정트리와 결정트리 기반의 앙상블 모형 중 랜덤포레스트 모형은 train set에서는 98%가 넘는 정확도로 매우 높게 나타나지만 test set에서는 각각 44.4%, 51.9%의 정확도로 매우 낮기 때문에 과대적합 되었다고 할 수 있다. 또한 결정트리 기반의 그래디언트 부스팅은 train set과 test set에서 모두 50% 대의 정확도로 낮은 정확성을 갖고 있다고 할 수 있다.

따라서 test set에서 앞에 10개, 뒤에 10개를 text와 author를 출력해 릿지 회귀를 통해 분류해본 결과 **총 20개의 소설 중 5개가 잘못 분류**되어 약 75%의 정확도를 보이는 것을 알 수 있다.

또한 author 별로 어떤 단어가 큰 영향을 주는지 상위 10개를 출력해본 결과, 다음과 같은 단어들의 영향이 크게 나타났다.

* **작가가 0**인 소설 : dora, darnay, joe, aunt, jerri, carton, handel, pip, creakl, crupp
* **작가가 1**인 소설 : ferrier, luci, lizzi, ann, shew, surpriz, tilney, jane, margaret, maria
* **작가가 2**인 소설 : inspector, professor, hordl, nigel, hudson, moor, bork, bowman, straker, pinner
* **작가가 3**인 소설 : princ, roubl, stepan, totski, moscow, ivan, suddenli, podincutor, simpli, zossimov
* **작가가 4**인 소설 : sila, casimir, villon, florizel, stubb, countess, silver, marjori, pool, balli